monai/networks/blocks/selfattention.py

```diff
@@ -34,46 +34,76 @@ def __init__(
         hidden_size: int,
         num_heads: int,
         dropout_rate: float = 0.0,
+        hidden_input_size: int | None = None,
+        dim_head: int | None = None,
         qkv_bias: bool = False,
         save_attn: bool = False,
+        causal: bool = False,
+        sequence_length: int | None = None,
         rel_pos_embedding: Optional[str] = None,
         input_size: Optional[Tuple] = None,
+        upcast_attention: bool = False,
     ) -> None:
         """
         Args:
             hidden_size (int): dimension of hidden layer.
             num_heads (int): number of attention heads.
             dropout_rate (float, optional): fraction of the input units to drop. Defaults to 0.0.
+            hidden_input_size (int, optional): dimension of the input tensor. Defaults to hidden_size.
+            dim_head (int, optional): dimension of each head. Defaults to hidden_size // num_heads.
```

Comment:
Looks like the change in this [PR](https://github.com/Project-MONAI/MONAI/pull/7664) already included in this one right?

Reply:
Yes unfortunately I realised that some of the changes in that PR were actually already existing in the Generative self-attention block and needed to be incorporated here in order to fully test everything here.

Apologies @NabJa for duplicating some work you had already done :(

Reply:
You said we should wait with my PR and then copied some of my code. This is not the way it should be done. You could have at least merged my working branch into yours. My changes are minor and would not have caused complex merge conflicts.

Reply:
Perhaps we can merge that one first since this refactor PR will be merged into `gen-ai-dev` branch, if there is any conflict we can solve it here?
Do you see any other concern for this proposal? @ericspod @marksgraham @NabJa

Reply:
Hi,

Yes my plan was always to ask @NabJa to merge their PR in, I'm happy to merge his in first and then solve conflicts here @KumoLiu .

@NabJa the reason I adopted much of the code from your PR was because I was hoping to reduce the merge conflicts when we did merge yours in. Maybe merging your branch in here would have been the better way to do it, I'll think about that next time :)